sihl/src/contract_queue.ml

```diff
@@ -1,64 +1,192 @@
+let default_retry_delay = Core_time.Span.minutes 1
+
+let default_error_handler id name input msg =
+  Lwt.return
+  @@ Logs.err (fun m ->
+         m
+           "Job with id '%s' and name '%s failed for input '%s': %s"
+           id
+           name
+           input
+           msg)
 ;;
 
-let pp fmt t = Sexplib0.Sexp.pp_hum fmt (to_sexp t)
-let default_tries = 5
-let default_retry_delay = Core_time.OneMinute
-
-let create ~name ~input_to_string ~string_to_input ~handle ?failed () =
-  let failed =
-    failed |> Option.value ~default:(fun _ -> Lwt_result.return ())
-  in
-  { name
-  ; input_to_string
-  ; string_to_input
-  ; handle
-  ; failed
-  ; max_tries = default_tries
-  ; retry_delay = default_retry_delay
-  }
+let create_job
+    handle
+    ?(max_tries = default_tries)
+    ?(retry_delay = default_retry_delay)
+    ?(failed = default_error_handler)
+    name
+  =
+  { name; handle; failed; max_tries; retry_delay }
 ;;
 
-let set_max_tries max_tries job = { job with max_tries }
-let set_retry_delay retry_delay job = { job with retry_delay }
+(* Service signature *)
+
+let name = "queue"
+
+module type Sig = sig
+  (** [router ?middlewares ?back scope] returns a router that can be passed to
+      the web server to serve the job queue dashboard.
+
+      [middlewares] is an optional list of middlewares that are appended to the
+      middlewares that the dashboard uses internally. Provide your own
+      authentication and authorization middlewares to make sure that only
+      privileged users can access the dashboard.
+
+      [back] is an optional URL which renders a back button on the dashboard.
+      Use this to provide your admin user a way to easily exit the dashboard. By
+      default, no URL is provided and no back button is shown.
+
+      [scope] is the URL path under which the dashboard can be accessed. It is
+      common to have some admin UI under [/admin], the job queue dashboard could
+      be available under [/admin/queue].
+
+      You can use HTMX by setting [HTMX_SCRIPT_URL] to the HTMX JavaScript file
+      that is embedded into the dashboard using the <script> tag in the page
+      body. HTMX is used to add dynamic features such as auto-refresh. The
+      dashboard is perfectly usable without it. By default, HTMX is not used. *)
+  val router
+    :  ?middlewares:Rock.Middleware.t list
+    -> ?back:string
+    -> string
+    -> Contract_http.router
+
+  (** [dispatch ?input ?delay job] queues [job] for later processing and returns
+      [unit Lwt.t] once the job has been queued.
+
+      [input] is the input of the [handle] function which is used for job
+      processing. By default, this is an empty string.
+
+      An optional [delay] determines the amount of time from now (when dispatch
+      is called) up until the job can be run. If no delay is specified, the job
+      is processed as soon as possible. *)
+  val dispatch : ?input:string -> ?delay:Ptime.span -> job -> unit Lwt.t
```

Comment:
Flip `input` and `delay` args please

Reply:
Fixed.

Reply:
Now the order of the explanation of the args in the docs is not the same as in `dispatch_all` (minor)

Reply:
Fixed.